packages/pirania/files/usr/libexec/rpcd/pirania

```diff
@@ -36,156 +41,103 @@ end
-        local randomKey = voucher.key
-        local newKey = key..'-'..randomKey
-        local output = { logic.add_voucher(db, newKey, voucher.voucher, voucher.epoc, 10, 10, 1) }
-        if (tonumber(output[1]) ~= 0 ) then
-            result.vouchers[_] = voucher.voucher
-        end
+local function renew_voucher(msg)
+    local new_expiration_date = tonumber(msg.date)
+    if new_expiration_date  then
+        local voucher = voucher.update_expiration_date(msg.id, new_expiration_date)
+        printJson({ success = voucher and 'ok' or 'error' })
     end
-    dba.save(config.db, db)
-    printJson(result)
 end
 
-local function add_member_voucher(msg)
-    local result = {}
-    local db = dba.load(config.db)
-    local key = utils.format_voucher_key(msg.note, 'member')
-    local secret = msg.secret
-    -- Should come from a shared-state json
-    local epoc = msg.epoc
-    local output = { logic.add_voucher(db, key, secret, epoc, 10, 10, 1) }
-    if (tonumber(output[0]) == 0) then
-        result.message = 'Error creating voucher'
-    else
-        result.node = hostname
-        result.key = key
-        result.type = 'member'
-        result.secret = secret
-        result.epoc = epoc
-        result.upload = 10
-        result.download = 10
-        result.amountofmacsallowed = 1
+local function add_vouchers(msg)
+    local vouchers = {}
+    for _, vouch_data in pairs (msg.vouchers_data) do
+        local v = {name=vouch_data.name, code=vouch_data.code, expiration_date=vouch_data.expiration_date}
+        local voucher, msg = vouchera.add(v)
+        if voucher == nil then
+            printJson({ status = 'error', message = msg})
+            return
+        end
+        table.insert(vouchers, voucher)
     end
-    dba.save(config.db, db)
-    printJson(result)
+    printJson({ status = 'ok', vouchers = vouchers})
 end
 
-local function auth_voucher(msg)
-    local result = {}
-    local mac = msg.mac
-    local voucher = msg.voucher
-    result.mac = mac
-    result.voucher = voucher
-    local db = dba.load(config.db)
-    local res = logic.auth_voucher(db, mac, voucher)
-    result.success = res.success
-    -- BUG: Saving to DB causes ubus to timeout
-    dba.save(config.db, db)
-    printJson(result)
+local function activate_voucher(msg)
+    local voucher = vouchera.activate(msg.code, msg.mac)
+    printJson({ success = voucher and 'ok' or 'error' })
 end
 
 local function print_valid_macs(msg)
     local result = {}
     result.macs = {}
-    local output = utils.shell('voucher print_valid_macs')
-    for line in output:gmatch("[^\n]+") do
-        -- local words = {}
-        -- for w in line:gmatch("%S+") do if w ~= "" then table.insert(words, w) end end
-        -- local mac = words[2]
-        table.insert(result.macs, line)
+    for _, voucher in pairs(vouchera.vouchers) do
+        if vouchera.is_active(voucher) then
+            table.insert(result.macs, voucher.mac)
+        end
     end
     printJson(result);
```

Comment:
have you seen this @spiccinini ?

Reply:
Yes, in all cases it may be refactored into a generic `apply_func_to_vouchers_if_condition(vouchers, func, condition)`, that is a `map(func, filter(condition, vouchers))` (I don't like a lot this generic functions, I prefer to do them explicit in code). Maybe a more specific  `apply_to_active_vouchers(function)` may be used (that I would prefer), but this won't serve all the cases mentioned. So for me it is not very clear that this refactor is a clear improvement (I mean maintainable without losing clarity and ease of reading).